### PR TITLE
pinocchio: 2.6.19 -> 2.6.20

### PR DESCRIPTION
**Pull request by nim65s**
## Description of changes

Upstream release for changelog: https://github.com/stack-of-tasks/pinocchio/releases/tag/v2.6.20

While here:
- add hpp-fcl dependency
- enable libpython dependency when python support is activated
- use finalAttrs instead of rec
- add check & python import check
- add myself in maintainers

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nim65s** [comment]
Result of `nixpkgs-review pr 248118` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.crocoddyl</li>
    <li>python310Packages.example-robot-data</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>crocoddyl</li>
    <li>pinocchio</li>
    <li>python310Packages.pinocchio</li>
    <li>python311Packages.pinocchio</li>
  </ul>
</details>

**Comment by nim65s** [comment]
Looks like I was over optimistic, and still have some debugging to do.

**Comment by nim65s** [comment]
Result of `nixpkgs-review pr 248118` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>crocoddyl</li>
    <li>example-robot-data</li>
    <li>pinocchio</li>
    <li>python310Packages.crocoddyl</li>
    <li>python310Packages.example-robot-data</li>
    <li>python310Packages.pinocchio</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
    <li>python311Packages.pinocchio</li>
  </ul>
</details>

File: pkgs/development/libraries/example-robot-data/default.nix [hunk 1]
```diff
@@ -33,6 +33,7 @@ stdenv.mkDerivation (finalAttrs: {
   ];
 
   doCheck = true;
+  preCheck = "ln -s /build/source /build/${finalAttrs.pname}";
```
[the remark this answers is not in the document]
Comment (nim65s) [reply pkgs/development/libraries/example-robot-data/default.nix:36]: Thanks for the review, I have no idea on how things work on darwin.

I'm testing another trick, without absolute paths.

[the remark this answers is not in the document]
Comment (nim65s) [reply pkgs/development/libraries/example-robot-data/default.nix]: `preCheck = "ln -s source ../../${finalAttrs.pname}";` looks too be working on linux. Will it be OK on darwin ?

[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/libraries/example-robot-data/default.nix]: `ln -s ../../source ../../${finalAttrs.pname}` should be OK on darwin.

[the remark this answers is not in the document]
Comment (nim65s) [reply pkgs/development/libraries/example-robot-data/default.nix]: ln syntax is different ? On linux this make a `../../example-robot-data` symlink to `../../../../source`:
```
nixbld@localhost:/var/lib/cntr/build/source/build$ file ../../example-robot-data       
../../example-robot-data: broken symbolic link to ../../source
nixbld@localhost:/var/lib/cntr/build/source/build$ ls ../../example-robot-data/robots
ls: cannot access '../../example-robot-data/robots': No such file or directory

File: pkgs/development/libraries/pinocchio/default.nix [hunk 2]
```diff
@@ -38,20 +39,31 @@ stdenv.mkDerivation rec {
...
-  cmakeFlags = lib.optionals (!pythonSupport) [
+  cmakeFlags = [
+    "-DBUILD_WITH_COLLISION_SUPPORT=ON"
+  ] ++ lib.optionals (pythonSupport) [
+    "-DBUILD_WITH_LIBPYTHON=ON"
+  ] ++ lib.optionals (!pythonSupport) [
     "-DBUILD_PYTHON_INTERFACE=OFF"
   ];
 
+  doCheck = true;
```
Comment (wegank) [inline pkgs/development/libraries/pinocchio/default.nix:57]:
```suggestion
  doCheck = !stdenv.isDarwin;
```

File: pkgs/development/libraries/example-robot-data/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/libraries/example-robot-data/default.nix]: I don't think this is the case. Doesn't `preCheck = "ls source";` give you an error under LInux?

File: pkgs/development/libraries/pinocchio/default.nix (hunk 2 above)
Comment (nim65s) [reply pkgs/development/libraries/pinocchio/default.nix]: Oh, wait, this is disabling tests on pinochio. Is this really what you intended ? I assumed it was for example-robot-data.

Comment (wegank) [reply pkgs/development/libraries/pinocchio/default.nix]: Yes, since five tests failed on ofborg. Haven't looked into it, though.

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 248118` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>crocoddyl</li>
    <li>python310Packages.crocoddyl</li>
    <li>python310Packages.example-robot-data</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>example-robot-data</li>
    <li>pinocchio</li>
    <li>python310Packages.pinocchio</li>
    <li>python311Packages.pinocchio</li>
  </ul>
</details>

File: pkgs/development/libraries/example-robot-data/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (nim65s) [reply pkgs/development/libraries/example-robot-data/default.nix]: then we can try:
```nix
preCheck = if stdenv.isDarwin
  then "ln -s source ../../${finalAttrs.pname}"
  else "ln -s ../../source ../../${finalAttrs.pname}";
  ```

[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/libraries/example-robot-data/default.nix]: There can't be any significant differences between the two `ln` on the two platforms. I mean, if `preCheck = "ls source"` doesn't work on Linux, then `preCheck = "ln -s source ../../${finalAttrs.pname}"` will only result in broken symbolic links.

**Comment by wegank** [comment]
It looks like the tests are broken in all reverse dependencies of pinocchio on darwin. Should we mark the package as broken there for now?

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 248118` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>crocoddyl</li>
    <li>python310Packages.crocoddyl</li>
    <li>python310Packages.example-robot-data</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>example-robot-data</li>
    <li>pinocchio</li>
    <li>python310Packages.pinocchio</li>
    <li>python311Packages.pinocchio</li>
  </ul>
</details>

File: pkgs/development/libraries/example-robot-data/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (nim65s) [reply pkgs/development/libraries/example-robot-data/default.nix]: Ok, therefore if `ln -s source ../../${finalAttrs.pname}` is working on linux, we should keep it, right ?

**Comment by nim65s** [comment]
Thanks. There is no hurry on my side.
If you can show me the logs you have on aarch64-darwin, I might be able to help.

**Comment by nim65s** [comment]
This is strange. I'll try to get a darwin setup in the future to dig further. 
But for this PR, your solution is totally fine for me, thanks !

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 248118` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>crocoddyl</li>
    <li>example-robot-data</li>
    <li>pinocchio</li>
    <li>python310Packages.crocoddyl</li>
    <li>python310Packages.example-robot-data</li>
    <li>python310Packages.pinocchio</li>
    <li>python311Packages.crocoddyl</li>
    <li>python311Packages.example-robot-data</li>
    <li>python311Packages.pinocchio</li>
  </ul>
</details>

**Pull request merged**